lib/fog/orchestration/util/recursive_hot_file_loader.rb

```diff
@@ -0,0 +1,250 @@
+
+          return true unless value.kind_of?(String)
+
+          return true if key == 'type' &&
+                         !value.end_with?('.yaml', '.template')
+
+          false
+        end
+
+        # Return true if I should inspect this yaml template branch.
+        def recurse_if(value)
+          value.kind_of?(Hash) || value.kind_of?(Array)
+        end
+
+        # Return string
+        def url_join(prefix, suffix)
+          if prefix
+            suffix = URI.join(prefix, suffix)
+            suffix = suffix.to_s.sub(%r{^file:\/+}, "file:///")
+          end
+          suffix
+        end
+
+        # Returns the string baseurl of the given url.
+        def base_url_for_url(url)
+          parsed = URI(url)
+          parsed_dir = File.dirname(parsed.path)
+          url_join(parsed, parsed_dir)
+        end
+
+        # Nothing to do on URIs
+        def normalise_file_path_to_url(path)
+          return path if URI(path).scheme
+
+          path = File.absolute_path(path)
+          url_join('file:/', path)
+        end
+
+        # Retrive the content of a local or remote file.
+        #
+        # @param A local or remote uri.
+        #
+        # @raise ArgumentError if it's not a valid uri
+        #
+        # Protect open-uri from malign arguments like
+        #  - "|ls"
+        #  - multiline strings
+        def get_content(uri_or_filename)
+          remote_schemes = %w(http https ftp)
+          Fog::Logger.debug("Opening #{uri_or_filename}")
+
+          begin
+            # Validate URI to protect from open-uri attacks.
+            url = URI(uri_or_filename)
+
+            # Remote schemes must contain an host.
+            raise ArgumentError if url.host.nil? && remote_schemes.include?(url.scheme)
+
+            # Encode URI with spaces.
+            uri_or_filename = URI.encode(URI.decode(URI(uri_or_filename).to_s))
+          rescue URI::InvalidURIError
+            raise ArgumentError, "Not a valid URI: #{uri_or_filename}"
+          end
+
+          # TODO: A future revision may implement a retry.
+          # TODO: A future revision may limit download size.
+          content = ''
+          # open-uri doesn't open "file:///" uris.
+          uri_or_filename = uri_or_filename.sub(%r/file:/, "")
+
+          open(uri_or_filename) { |f| content = f.read }
+          content
+        end
+
+        # Retrieve a template content.
+        #
+        # @param template_file can be either:
+        #          - a raw_template string
+        #          - an URI string
+        #          - an Hash containing the parsed template.
+        #
+        # XXX: after deprecation of Ruby 1.9 we could use
+        #      named parameters and better mimic heatclient implementation.
+        def get_template_contents(template_file)
+          Fog::Logger.warning("get_template_contents #{template_file}")
+
+          raise "template_file should be Hash or String" unless
```

Comment:
This is private method for internal use. We may not need to validate the argument all the time.

Reply:
@bzwei  `get_template_file` receives the template directly from `get_files`.

Validating here make the code safer and allows to reuse this method for `resource_registry` & co.